app/services/promo/registrations_geo_stats_fetcher.rb

```diff
@@ -1,29 +1,71 @@
 # Fetches and does not save geo stats for a list of referral_codes
 class Promo::RegistrationsGeoStatsFetcher < BaseApiClient
   include PromosHelper
-  BATCH_SIZE = 50
+  BATCH_SIZE = 10
+  MAX_RETRY = 3
 
-  def initialize(promo_registrations:)
-    @referral_codes = promo_registrations.map { |promo_registration| promo_registration.referral_code }
+  def initialize(referral_codes:, start_date: nil, end_date: nil, interval: nil)
+    @referral_codes = referral_codes
+    @start_date = start_date
+    @end_date = end_date
+    @interval =
+      case interval
+      when PromoRegistration::DAILY
+        'day'
+      when PromoRegistration::WEEKLY
+        'week'
+      when PromoRegistration::MONTHLY
+        'month'
+      end
   end
 
   def perform
+    params = {
+      start_date: @start_date,
+      end_date: @end_date,
+      interval: @interval,
+    }
+
     return perform_offline if Rails.application.secrets[:api_promo_base_uri].blank?
+
     geo_stats = []
-    @referral_codes.each_slice(BATCH_SIZE).to_a.each do |referral_code_batch|
-      query_string = query_string(referral_code_batch)
-      response = connection.get do |request|
-        request.options.params_encoder = Faraday::FlatParamsEncoder
-        request.headers["Authorization"] = api_authorization_header
-        request.headers["Content-Type"] = "application/json"
-        request.url("/api/2/promo/geoStatsByReferralCode#{query_string}")
+    @referral_codes.each_slice(BATCH_SIZE).to_a.each do |batch|
+      response = retry_request(MAX_RETRY) do
+        request({ referral_code: batch }.merge(params))
       end
-      geo_stats = geo_stats + JSON.parse(response.body)
+
+      raise StandardError.new(message: 'Could not fetch geo stats due to an error from Promo Server') if response.blank?
+
+      geo_stats += JSON.parse(response.body)
     end
+
     geo_stats
+  end
+
+  def request(params)
+    connection.get do |request|
+      request.options.params_encoder = Faraday::FlatParamsEncoder
+      request.headers["Authorization"] = api_authorization_header
+      request.headers["Content-Type"] = "application/json"
+      request.params = params.compact
+      request.url("/api/2/promo/geoStatsByReferralCode")
+    end
   rescue Faraday::ClientError => e
     Rails.logger.info("Error receiving referral stats #{e.message}")
-    []
+    nil
```

Comment:
Would it be useful to log this in Sentry?

Reply:
I don't think so, only because these errors are fairly consistent. The StandardError will be shown in Sentry if all the HTTP responses fail and then the engineer responsible can search the logs for the corresponding message.